the latest gonum library using Nodes instead of []Node

..\github.com\chewxy\gorgonia\walker.go:43:43: cannot use g (type *ExprGraph) as type graph.Directed in argument to topo.SortStabilized:
        *ExprGraph does not implement graph.Directed (wrong type for From method)
                have From(int64) []graph.Node
                want From(int64) graph.Nodes

--- a/compile.go
+++ b/compile.go
@@ -18,7 +18,7 @@ func Compile(g *ExprGraph) (prog *program, locMap map[*Node]register, err error)
 	enterLogScope()
 	defer leaveLogScope()
 
-	if len(g.Nodes()) == 0 {
+	if g.Nodes().Len() == 0 {
 		err = errors.Errorf("Cannot compile an empty graph")
 		return
 	}

--- a/const.go
+++ b/const.go
@@ -34,7 +34,7 @@ const (
 
 	// error messages
 	sortFail            = "Failed to sort"
-	cloneFail           = "Failed to carry clone()"
+	cloneFail           = "Failed to carry clone(%v)"
 	clone0Fail          = "Failed to carry clone0()"
 	nyiTypeFail         = "%s not yet implemented for %T"
 	nyiFail             = "%s not yet implemented for %v"

--- a/differentiation.go
+++ b/differentiation.go
@@ -1,6 +1,9 @@
 package gorgonia
 
-import "github.com/pkg/errors"
+import (
+	"github.com/pkg/errors"
+	"gonum.org/v1/gonum/graph"
+)
 
 /*
 This file holds code for symbolic differentiation.
@@ -100,7 +103,7 @@ func backwardDiffAnalysis(wrt, sortedNodes Nodes) (retVal NodeSet, err error) {
 			}
 			g := n.g
 			for _, child := range n.children {
-				parents := g.To(child.ID())
+				parents := graph.NodesOf(g.To(child.ID()))
 
 				symdiffLogf("parents of %v: %v", child, graphNodeToNode(parents))
 				if len(parents) == 1 && len(child.children) > 0 {
@@ -156,8 +159,8 @@ func Backpropagate(outputs, gradOutputs, wrt Nodes) (retVal Nodes, err error) {
 	for i := 0; i < len(g.AllNodes()); i++ {
 		n := g.AllNodes()[i]
 
-		fr := len(g.From(n.ID()))
-		to := len(g.To(n.ID()))
+		fr := g.From(n.ID()).Len()
+		to := g.To(n.ID()).Len()
 
 		if fr == 0 && to == 0 && !n.isConstant() && !n.isInput() {
 			g.RemoveNode(n)

--- a/graph.go
+++ b/graph.go
@@ -6,6 +6,7 @@ import (
 
 	"github.com/awalterschulze/gographviz"
 	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/iterator"
 )
 
 // ExprGraph is a data structure for a directed acyclic graph (of expressions). This structure is the main entry point
@@ -520,12 +521,12 @@ func (g *ExprGraph) Has(nodeid int64) bool {
 }
 
 // Nodes returns all the nodes in the graph.
-func (g *ExprGraph) Nodes() []graph.Node {
+func (g *ExprGraph) Nodes() graph.Nodes {
 	// nodes := make([]graph.Node, len(g.from))
 	ns := g.AllNodes()
 
 	nodes := nodeToGraphNode(ns)
-	return nodes
+	return iterator.NewOrderedNodes(nodes)
 }
 
 // AllNodes is like Nodes, but returns Nodes instead of []graph.Node.
@@ -533,9 +534,9 @@ func (g *ExprGraph) Nodes() []graph.Node {
 func (g *ExprGraph) AllNodes() Nodes { return g.all }
 
 // From returns all nodes in g that can be reached directly from n.
-func (g *ExprGraph) From(nodeid int64) []graph.Node {
+func (g *ExprGraph) From(nodeid int64) graph.Nodes {
 	if n := g.node(nodeid); n != nil {
-		return nodeToGraphNode(n.children)
+		return iterator.NewOrderedNodes(nodeToGraphNode(n.children))
 	}
 	return nil
 }
@@ -583,7 +584,7 @@ func (g *ExprGraph) HasEdgeFromTo(u, v int64) bool {
 }
 
 // To returns all nodes in g that can reach directly to n.
-func (g *ExprGraph) To(nid int64) []graph.Node {
+func (g *ExprGraph) To(nid int64) graph.Nodes {
 	n := g.node(nid)
 	if n == nil {
 		return nil
@@ -592,7 +593,7 @@ func (g *ExprGraph) To(nid int64) []graph.Node {
 	ns := g.to[n]
 	ns = ns.Set()
 	g.to[n] = ns
-	return nodeToGraphNode(ns)
+	return iterator.NewOrderedNodes(nodeToGraphNode(ns))
 }
 
 // subgraph is basically a subset of nodes. This is useful for compiling sub sections of the graph

--- a/graph_test.go
+++ b/graph_test.go
@@ -45,10 +45,10 @@ func TestGraphBasics(t *testing.T) {
 	assert.Equal(correctTo, g.to[x])
 
 	correctTo = Nodes{xy}
-	assert.Equal(correctTo, graphNodeToNode(g.To(y.ID())))
-	assert.Equal(correctTo, graphNodeToNode(g.To(x.ID())))
+	assert.Equal(correctTo, graphNodeToNode(graph.NodesOf(g.To(y.ID()))))
+	assert.Equal(correctTo, graphNodeToNode(graph.NodesOf(g.To(x.ID()))))
 
-	assert.Equal(3, len(g.Nodes()))
+	assert.Equal(3, g.Nodes().Len())
 
 	// Now, time to deal with constants
 	xy1 := Must(Add(xy, onef64))
@@ -56,7 +56,7 @@ func TestGraphBasics(t *testing.T) {
 	assert.Equal(g, xy1.g)
 
 	var containsOne bool
-	for _, node := range g.Nodes() {
+	for _, node := range graph.NodesOf(g.Nodes()) {
 		n := node.(*Node)
 		if n.Hashcode() == onef64.Hashcode() {
 			containsOne = true

--- a/node.go
+++ b/node.go
@@ -11,6 +11,7 @@ import (
 	"github.com/awalterschulze/gographviz"
 	"github.com/chewxy/hm"
 	"github.com/pkg/errors"
+	"gonum.org/v1/gonum/graph"
 	"gorgonia.org/tensor"
 )
 
@@ -537,7 +538,7 @@ func (n *Node) RestrictedToDot(up, down int) string {
 		origLen := len(upQ)
 		for i := 0; i < origLen; i++ {
 			qn := upQ[i]
-			toQN := graphNodeToNode(g.To(qn.ID()))
+			toQN := graphNodeToNode(graph.NodesOf(g.To(qn.ID())))
 			upQ = append(upQ, toQN...)
 			ns = append(ns, toQN...)
 		}

--- a/regalloc.go
+++ b/regalloc.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 
 	"github.com/xtgo/set"
+	"gonum.org/v1/gonum/graph"
 )
 
 // this file holds all the code that relates to register allocation
@@ -184,7 +185,7 @@ func (ra *regalloc) allocMutableOp(node *Node, nInterv *interval) {
 	compileLogf("Read %v", reads)
 
 	var letStmts Nodes
-	for _, parent := range node.g.To(node.ID()) {
+	for _, parent := range graph.NodesOf(node.g.To(node.ID())) {
 		n := parent.(*Node)
 		compileLogf("Parent: %v | %T", n, n.op)
 		if n.isStmt {

